components/tools/OmeroWeb/test/integration/test_api_wells.py

```diff
@@ -0,0 +1,231 @@
+
+
+def get_connection(user, group_id=None):
+    """Get a BlitzGateway connection for the given user's client."""
+    connection = BlitzGateway(client_obj=user[0])
+    # Refresh the session context
+    connection.getEventContext()
+    if group_id is not None:
+        connection.SERVICE_OPTS.setOmeroGroup(group_id)
+    return connection
+
+
+def cmp_name_insensitive(x, y):
+    """Case-insensitive name comparator."""
+    return cmp(unwrap(x.name).lower(), unwrap(y.name).lower())
+
+
+def cmp_column_row(x, y):
+    """Sort wells by row, then column."""
+    sort_by_column = cmp(unwrap(x.column), unwrap(y.column))
+    if sort_by_column == 0:
+        return cmp(unwrap(x.row), unwrap(y.row))
+    return sort_by_column
+
+
+def marshal_objects(objects):
+    """Marshal objects using omero_marshal."""
+    expected = []
+    for obj in objects:
+        encoder = get_encoder(obj.__class__)
+        expected.append(encoder.encode(obj))
+    return expected
+
+
+def remove_urls(marshalled):
+    """Traverse a dict (Well) removing 'url:' values."""
+    for key, val in marshalled.items():
+        if key.startswith('url:'):
+            del(marshalled[key])
+        # We only traverse paths where we know urls are
+        elif key == 'Image':        # isinstance(val, dict):
+            remove_urls(val)
+        elif key == 'WellSamples':  # isinstance(val, list):
+            for i in val:
+                remove_urls(i)
+
+
+def assert_objects(conn, json_objects, omero_ids_objects, dtype="Project",
+                   group='-1', extra=None, opts=None):
+    """
+    Load objects from OMERO, via conn.getObjects().
+
+    marshal with omero_marshal and compare with json_objects.
+    omero_ids_objects can be IDs or list of omero.model objects.
+
+    @param: extra       List of dicts containing expected extra json data
+                        e.g. {'omero:childCount': 1}
+    """
+    pids = []
+    for p in omero_ids_objects:
+        try:
+            pids.append(long(p))
+        except TypeError:
+            pids.append(p.id.val)
+    conn.SERVICE_OPTS.setOmeroGroup(group)
+    objs = conn.getObjects(dtype, pids, respect_order=True, opts=opts)
+    objs = [p._obj for p in objs]
+    expected = marshal_objects(objs)
+    assert len(json_objects) == len(expected)
+    for i, o1, o2 in zip(range(len(expected)), json_objects, expected):
+        if extra is not None and i < len(extra):
+            o2.update(extra[i])
+        # dumping to json and loading (same as test data) means that
+        # unicode has been handled in same way, e.g. Pixel size symbols.
+        o2 = json.loads(json.dumps(o2))
+        # remove any urls from json (tested elsewhere)
+        remove_urls(o1)
+
+        assert o1 == o2
+
+
+class TestWells(IWebTest):
+    """Tests querying Wells."""
+
+    @pytest.fixture()
+    def user1(self):
+        """Return a new user in a read-annotate group."""
+        group = self.new_group(perms='rwra--')
```

Comment:
I'm not sure what you're wanting here. Testing that queries still work in a read-only or private group? Tests in ```test_api_projects.py``` do various permissions & cross-group queries so I don't think we need to replicate this in other tests that are focussed on different objects?